leader-election-impl/src/main/java/com/palantir/paxos/LeaderPingerContext.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.paxos;
+
+import org.immutables.value.Value;
+
+import com.google.common.net.HostAndPort;
+
+@Value.Immutable
+@Value.Style(allParameters = true)
+public interface LeaderPingerContext<T> {
```

Comment:
Do you need the generic param here?

Reply:
This looks like a lot of typing all around the place, maybe we could restructure SingleLeaderPinger somehow, so maybe it just uses HostAndPort everywhere as keys, instead of this thing.

Reply:
it's not just `SingleLeaderPinger` that uses this. Granted, could probably just make two separate immutable objects that contain `PingableLeader` and `HostAndPort` and `BatchPingableLeader` and `HostAndPort`.

The issue here is that we need to keep the association between the rpc class we just invoked ping on and the the host and port information. It's easier to keep them as a "tuple" as opposed to having another map (in both impls) to associate the rpc class and the address info.

In an ideal world with retrofit, this would be pretty straightforward since we'd just get the information from `Call<T>`. But alas, we are not there.

Reply:
I feel like HostAndPort is the unique key though? So maybe the BatchPingableLeader and PingableLeader should be part of the value, not key?

Reply:
It is unique, but not useful by itself. The `PaxosQuorumChecker` takes in `Map<RPC, Executor>`. If you want to just have a single key (`RPC`), you will need a separate map to keep the association between `HostAndPort` and `RPC` in all the places you will use this.